packages/react-components/src/templates/TagsPageBody.tsx

```diff
@@ -29,8 +41,44 @@ const MessageBody: React.FC<{ title: string; body: string }> = ({
   </main>
 );
 
+export enum TagFieldByEntity {
+  'research-output' = 'keywords',
+  user = 'expertiseAndResourceTags',
+  event = 'calendar',
+}
+
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+const assert = <T extends never>() => undefined;
+type TypeEqualityGuard<A, B> = Exclude<A, B> | Exclude<B, A>;
+
+const EntityCard: React.FC<TagsPageBodyProps['results'][number]> = ({
+  ...data
+}): JSX.Element => {
+  assert<
+    TypeEqualityGuard<keyof typeof TagFieldByEntity, CRNTagSearchEntities>
+  >;
+
+  if (TagFieldByEntity['research-output'] in data) {
+    return <SharedResearchCard {...data} />;
+  }
+
+  if (TagFieldByEntity.user in data) {
+    return <PeopleCard {...data} />;
+  }
+
+  if (TagFieldByEntity.event in data) {
+    return <EventCard {...eventMapper(data)} />;
+  }
+
+  return <ErrorCard />;
```

Comment:
Do we need this case? Shouldn't the types prevent this from happening?

Reply:
You're right! This isn't necessary, I've removed it here [review: remove unnecessary error card](https://github.com/yldio/asap-hub/pull/3856/commits/674e1b8e74b778d5051ef9e62dcb2f5be7bbe1d1)